jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java

```diff
@@ -1190,98 +1170,76 @@ public boolean isReady()
                     return true;
 
                 case ASYNC:
-                    if (!_state.compareAndSet(State.ASYNC, State.READY))
-                        continue;
+                    _state = State.READY;
                     return true;
 
                 case PENDING:
-                    if (!_state.compareAndSet(State.PENDING, State.UNREADY))
-                        continue;
+                    _state = State.UNREADY;
                     return false;
 
                 case UNREADY:
                     return false;
 
                 default:
-                    throw new IllegalStateException();
+                    throw new IllegalStateException(_state.toString());
             }
         }
     }
 
     @Override
     public void run()
     {
-        while (true)
+        Throwable error = null;
+
+        synchronized (_channelState)
         {
-            State state = _state.get();
+            if (_state == State.ERROR)
+            {
+                error = _onError;
+                _onError = null;
+            }
+        }
 
-            if (_onError != null)
+        try
+        {
+            if (error == null)
             {
-                switch (state)
-                {
-                    case CLOSING:
-                    case CLOSED:
-                    case ERROR:
-                    {
-                        _onError = null;
-                        return;
-                    }
-                    default:
-                    {
-                        if (_state.compareAndSet(state, State.ERROR))
-                        {
-                            Throwable th = _onError;
-                            _onError = null;
-                            if (LOG.isDebugEnabled())
-                                LOG.debug("onError", th);
-
-                            try
-                            {
-                                _writeListener.onError(th);
-                            }
-                            finally
-                            {
-                                IO.close(this);
-                            }
-
-                            return;
-                        }
-                    }
-                }
-                continue;
+                if (LOG.isDebugEnabled())
+                    LOG.debug("onWritePossible");
+                _writeListener.onWritePossible();
```

Comment:
If we are in state `CLOSED` or `CLOSING` we will now call `onWritePossible()` here?

Reply:
I think we want to do that.... if we have got to UNREADY state we should call owp even if we are closed.
But I'm only 95% sure of that...

There are strange situations... like if we are closed then isReady should return true (else another dispatch is scheduled)... 
but if we sent content length and then the very last write of the content congests for a while - what should isReady return?  true because it knows that the connection will be closed once the write is complete, or false thus causing an OWP callback with a closed output!    I think the later is correct... but we need to review carefully!